### PR TITLE
Fix bug with spinner for vitamin_a in ProductEditNutritionFactsFragment

### PR DESCRIPTION
**Pull request by naivekook**
####  Description
Found a bug for "vitamin_a" nutrinion in ProductEditNutritionFactsFragment.kt when we try to add row
```
2021-02-08 19:01:00.312 869-869/openfoodfacts.github.scrachx.openfood.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: openfoodfacts.github.scrachx.openfood.debug, PID: 869
    java.lang.ArrayIndexOutOfBoundsException: length=3; index=4
        at java.util.Arrays$ArrayList.get(Arrays.java:3769)
        at android.widget.ArrayAdapter.getItem(ArrayAdapter.java:393)
        at android.widget.ArrayAdapter.createViewFromResource(ArrayAdapter.java:449)
        at android.widget.ArrayAdapter.getView(ArrayAdapter.java:415)
        at android.widget.Spinner.makeView(Spinner.java:712)
        at android.widget.Spinner.layout(Spinner.java:660)
        at android.widget.Spinner.onLayout(Spinner.java:622)
```

#### Related issues
not found

#### Related PRs
not found

#### Screenshots
|Screenshot_1613045536

### CODE REVIEW

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt [hunk 1]
```diff
@@ -693,19 +693,27 @@ class ProductEditNutritionFactsFragment : ProductEditFragment() {
...
+            Nutriments.PH -> {
+                unitSpinner.visibility = View.INVISIBLE
+            }
+            Nutriments.STARCH -> {
+                val arrayAdapter = ArrayAdapter(requireActivity(), android.R.layout.simple_spinner_item, requireActivity().resources.getStringArray(R.array.weights_array))
+                arrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
+                unitSpinner.adapter = arrayAdapter
+                starchEditText = editText
+            }
+            Nutriments.VITAMIN_A -> {
```
Comment (naivekook) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: looks like we have to setup IU measurement for all vitamins, not only for "A" right?

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt [hunk 2]
```diff
@@ -693,19 +693,27 @@ class ProductEditNutritionFactsFragment : ProductEditFragment() {
         editText.imeOptions = EditorInfo.IME_ACTION_DONE
         editText.requestFocus()
         if (preFillValues) {
-            editText.setText(value)
+            value?.takeIf { it != "0" }?.let { editText.setText(value) }
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: I think "0" is an acceptable value :)

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt (hunk 1 above)
Comment (VaiTon) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: I think we should. Could you try to see if other vitamins have the same problem?

Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: yes will try

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt (hunk 2 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: ah ok, will revert it

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt (hunk 1 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: found in the internet that the IU usually used to measure fat soluble vitamins including Vitamin A, D and E, so will make this change for these three vitamins

**Comment by naivekook** [comment]
@VaiTon please take a look again

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt [hunk 3]
```diff
@@ -647,16 +640,18 @@ class ProductEditNutritionFactsFragment : ProductEditFragment() {
         }
 
     private fun displayAddNutrientDialog() {
-        val origNutrients = resources.getStringArray(R.array.nutrients_array).toMutableList()
-        usedNutrientsIndexes.forEach { origNutrients.removeAt(it) }
```
Comment (naivekook) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt:650]: there was an IndexOutOfBoundsException

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt [hunk 4]
```diff
@@ -157,8 +155,8 @@ class ProductEditNutritionFactsFragment : ProductEditFragment() {
             it.addValidListener()
             it.checkValue()
         }
-        if (getActivity() is ProductEditActivity && (getActivity() as ProductEditActivity?)!!.initialValues != null) {
-            addAllFieldsToMap((getActivity() as ProductEditActivity?)!!.initialValues!!)
+        (requireActivity() as? ProductEditActivity)?.initialValues?.let { values ->
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]:
```suggestion
        (getActivity() as? ProductEditActivity)?.initialValues?.let { values ->
```

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt [hunk 5]
```diff
@@ -699,22 +695,41 @@ class ProductEditNutritionFactsFragment : ProductEditFragment() {
...
-        if (preFillValues) {
-            unitSpinner.setSelection(unitSelectedIndex)
-            modSpinner.setSelection(modSelectedIndex)
+        try {
+            if (preFillValues) {
+                unitSpinner.setSelection(unitSelectedIndex)
+                modSpinner.setSelection(modSelectedIndex)
+            }
+        } catch (t: Throwable) {
+            AnalyticsService.record(Throwable("Can't find weight units for nutriment: $nutrientShortName", t))
```
Comment (VaiTon) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]:
```suggestion
            AnalyticsService.record(IllegalStateException("Can't find weight units for nutriment: $nutrientShortName", t))
```

Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: done

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt (hunk 4 above)
Comment (naivekook) [reply app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/edit/ProductEditNutritionFactsFragment.kt]: done

**Review by VaiTon — APPROVED** [review]

**Comment by naivekook** [comment]
@teolemon PR is ready to go

**Pull request merged**